src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -1861,156 +1819,155 @@ final void registerWorker(WorkQueue w) {
      * @param ex the exception causing failure, or null if none
      */
     final void deregisterWorker(ForkJoinWorkerThread wt, Throwable ex) {
+        if ((runState & STOP) != 0L)       // ensure released
+            releaseAll();
         WorkQueue w = null;
         int src = 0, phase = 0;
         boolean replaceable = false;
         if (wt != null && (w = wt.workQueue) != null) {
             phase = w.phase;
-            if ((src = w.source) != DEREGISTERED) { // else trimmed on timeout
-                w.source = DEREGISTERED;
-                if (phase != 0) {         // else failed to start
+            if ((src = w.source) != DROPPED) {
+                w.source = DROPPED;        // else already dropped
+                if (phase != 0) {          // else failed to start
                     replaceable = true;
-                    if ((phase & IDLE) != 0)
-                        reactivate(w);    // pool stopped before released
+                    if (w.top - w.base > 0) {
```

Comment:
@DougLea Is this check sufficiently cheaper than trying to poll blind?

Reply:
Done.